sb3_contrib/crossq/crossq.py

```diff
@@ -0,0 +1,336 @@
+            self.target_entropy = float(self.target_entropy)
+
+        # The entropy coefficient or entropy can be learned automatically
+        # see Automating Entropy Adjustment for Maximum Entropy RL section
+        # of https://arxiv.org/abs/1812.05905
+        if isinstance(self.ent_coef, str) and self.ent_coef.startswith("auto"):
+            # Default initial value of ent_coef when learned
+            init_value = 1.0
+            if "_" in self.ent_coef:
+                init_value = float(self.ent_coef.split("_")[1])
+                assert init_value > 0.0, "The initial value of ent_coef must be greater than 0"
+
+            # Note: we optimize the log of the entropy coeff which is slightly different from the paper
+            # as discussed in https://github.com/rail-berkeley/softlearning/issues/37
+            self.log_ent_coef = th.log(th.ones(1, device=self.device) * init_value).requires_grad_(True)
+            self.ent_coef_optimizer = th.optim.Adam([self.log_ent_coef], lr=self.lr_schedule(1))
+        else:
+            # Force conversion to float
+            # this will throw an error if a malformed string (different from 'auto')
+            # is passed
+            self.ent_coef_tensor = th.tensor(float(self.ent_coef), device=self.device)
+
+    def _create_aliases(self) -> None:
+        self.actor = self.policy.actor
+        self.critic = self.policy.critic
+
+    def train(self, gradient_steps: int, batch_size: int = 64) -> None:
+        # Switch to train mode (this affects batch norm / dropout)
+        self.policy.set_training_mode(True)
+        # Update optimizers learning rate
+        optimizers = [self.actor.optimizer, self.critic.optimizer]
+        if self.ent_coef_optimizer is not None:
+            optimizers += [self.ent_coef_optimizer]
+
+        # Update learning rate according to lr schedule
+        self._update_learning_rate(optimizers)
+
+        ent_coef_losses, ent_coefs = [], []
+        actor_losses, critic_losses = [], []
+
+        for _ in range(gradient_steps):
+            self._n_updates += 1
+            # delayed updates
+            update_actor_and_temperature = self._n_updates % self.policy_delay == 0
+            # Sample replay buffer
+            replay_data = self.replay_buffer.sample(batch_size, env=self._vec_normalize_env)  # type: ignore[union-attr]
+
+            # We need to sample because `log_std` may have changed between two gradient steps
+            if self.use_sde:
+                self.actor.reset_noise()
+
+            # Note: in the following lines we always need to make sure to set train/eval modes
+            # of actor and critic carefully. This is because of the BatchNorm layers in the networks
+            # which behave differently in train and eval modes.
+
+            self.actor.set_training_mode(True)
+            actions_pi, log_prob = self.actor.action_log_prob(replay_data.observations)
```

Comment:
minor but still a difference with original implementation: the batch norm stats for the actor are updated at every training step?
and the ent coef is updated at every training step too?
(in other word, the delay parameter is not taken into account?)

Reply:
good catch, that's true. I guess the easiest fix for this would be to set `self.actor.set_training_mode(update_actor_and_temperature)`.
What do you think?

Reply:
i reorganize things and now it should be as in the original implementation